File: src/shady/generator/generator_common.c
#include "generator.h"

bool has_custom_ctor(json_object* node) {
    String constructor = json_object_get_string(json_object_object_get(node, "constructor"));
    return (constructor && strcmp(constructor, "custom") == 0);
}

json_object* lookup_node_class(json_object* src, String name) {
    json_object* node_classes = json_object_object_get(src, "node-classes");
    for (size_t i = 0; i < json_object_array_length(node_classes); i++) {
        json_object* class = json_object_array_get_idx(node_classes, i);
        String class_name = json_object_get_string(json_object_object_get(class, "name"));
        assert(class_name);
        if (strcmp(name, class_name) == 0)
            return class;
    }
    return NULL;
}

String class_to_type(json_object* src, String class, bool list) {
    assert(class);
    if (strcmp(class, "string") == 0) {
        if (list)
            return "Strings";
        else
            return "String";
    }
    // check the class is valid
    if (!lookup_node_class(src, class)) {
        shd_error_print("invalid node class '%s'\n", class);
        shd_error_die();
    }
    return list ? "Nodes" : "const Node*";
}

bool find_in_set(json_object* node, String class_name) {
    switch (json_object_get_type(node)) {
        case json_type_array: {
            for (size_t i = 0; i < json_object_array_length(node); i++)
                if (find_in_set(json_object_array_get_idx(node, i), class_name))
                    return true;
            break;
        }
        case json_type_string: return strcmp(json_object_get_string(node), class_name) == 0;
        default: break;
    }
    return false;
}

String get_type_for_operand(json_object* src, json_object* op) {
    String op_type = json_object_get_string(json_object_object_get(op, "type"));
    bool list = json_object_get_boolean(json_object_object_get(op, "list"));
    String op_class = NULL;
    if (!op_type) {
        op_class = json_object_get_string(json_object_object_get(op, "class"));
        op_type = class_to_type(src, op_class, list);
    }
    assert(op_type);
    return op_type;
}

void preprocess(json_object* src) {
    json_object* nodes = json_object_object_get(src, "nodes");
    for (size_t i = 0; i < json_object_array_length(nodes); i++) {
        json_object* node = json_object_array_get_idx(nodes, i);
        String name = json_object_get_string(json_object_object_get(node, "name"));
        json_object* snake_name = json_object_object_get(node, "snake_name");
        if (!snake_name) {
            String tmp = to_snake_case(name);
            json_object* generated_snake_name = json_object_new_string(tmp);
            json_object_object_add(node, "snake_name", generated_snake_name);
            free((void*) tmp);
        }
    }
}

void generate_bit_enum(Growy* g, String enum_type_name, String enum_case_prefix, json_object* cases) {
    assert(json_object_get_type(cases) == json_type_array);
    shd_growy_append_formatted(g, "typedef enum {\n");
    for (size_t i = 0; i < json_object_array_length(cases); i++) {
        json_object* node_class = json_object_array_get_idx(cases, i);
        String name = json_object_get_string(json_object_object_get(node_class, "name"));
        String capitalized = capitalize(name);
        shd_growy_append_formatted(g, "\t%s%s = 0x%x", enum_case_prefix, capitalized, (1 << i));
        shd_growy_append_formatted(g, ",\n");
        free((void*) capitalized);
    }
    shd_growy_append_formatted(g, "} %s;\n\n", enum_type_name);
}

void generate_bit_enum_classifier(Growy* g, String fn_name, String enum_type_name, String enum_case_prefix, String src_type_name, String src_case_prefix, String src_case_suffix, json_object* cases) {
    shd_growy_append_formatted(g, "%s %s(%s tag) {\n", enum_type_name, fn_name, src_type_name);
    shd_growy_append_formatted(g, "\tswitch (tag) { \n");
    assert(json_object_get_type(cases) == json_type_array);
    for (size_t i = 0; i < json_object_array_length(cases); i++) {
        json_object* node = json_object_array_get_idx(cases, i);
        String name = json_object_get_string(json_object_object_get(node, "name"));
        shd_growy_append_formatted(g, "\t\tcase %s%s%s: \n", src_case_prefix, name, src_case_suffix);
        json_object* class = json_object_object_get(node, "class");
        switch (json_object_get_type(class)) {
            case json_type_null:shd_growy_append_formatted(g, "\t\t\treturn 0;\n");
                break;
            case json_type_string: {
                String cap = capitalize(json_object_get_string(class));
                shd_growy_append_formatted(g, "\t\t\treturn %s%s;\n", enum_case_prefix, cap);
                free((void*) cap);
                break;
            }
            case json_type_array: {
                shd_growy_append_formatted(g, "\t\t\treturn ");
                for (size_t j = 0; j < json_object_array_length(class); j++) {
                    if (j > 0)
                        shd_growy_append_formatted(g, " | ");
                    String cap = capitalize(json_object_get_string(json_object_array_get_idx(class, j)));
                    shd_growy_append_formatted(g, "%s%s", enum_case_prefix, cap);
                    free((void*) cap);
                }
                shd_growy_append_formatted(g, ";\n");
                break;
            }
            case json_type_boolean:
            case json_type_double:
            case json_type_int:
            case json_type_object:
                shd_error_print("Invalid datatype for a node's 'class' attribute");
                break;
        }
    }
    shd_growy_append_formatted(g, "\t\tdefault: assert(false);\n");
    shd_growy_append_formatted(g, "\t}\n");
    shd_growy_append_formatted(g, "\tSHADY_UNREACHABLE;\n");
    shd_growy_append_formatted(g, "}\n");
}

File: src/generator/generator_common.py
from __future__ import absolute_import
from . generator import *

def is_recursive_node(node):
    return node.get("recursive") == True

def lookup_node_class(src, name):
    for c in src["node-classes"]:
        if c["name"] == name:
            return c
    return None

def class_to_type(src, clazz, list):
    assert clazz
    if clazz == "string":
        return "Strings" if list else "String"
    if clazz != "none" and lookup_node_class(src, clazz) is None:
        print(f"Invalid node class '{clazz}'")
    return "Nodes" if list else "const Node*"

def find_in_set(set_or_value, item):
    if set_or_value == item:
        return True
    if item in set_or_value:
        return True
    return False

def get_type_for_operand(src, op):
    op_type = op.get("type")
    list = op.get("list")
    if op_type is None:
        op_class = op["class"]
        assert op_class
        op_type = class_to_type(src, op_class, list)
    return op_type

def preprocess(src):
    for node in src["nodes"]:
        if node.get("snake_name") is None:
            node["snake_name"] = to_snake_case(node["name"])

def generate_bit_enum(g, enum_type_name, enum_case_prefix, cases, include_none):
    assert type(cases) is list
    g.g += "typedef enum {\n"
    if include_none:
        g.g += f"\t{enum_case_prefix}None = 0,\n"
    i = 1
    for node_class in cases:
        name = node_class["name"]
        capitalized = capitalize(name)
        g.g += f"\t{enum_case_prefix}{capitalized} = {hex(i)},\n"
        i *= 2
    g.g += "}" + f" {enum_type_name};\n\n"

def generate_bit_enum_classifier(g, fn_name, enum_type_name, enum_case_prefix, src_type_name, src_case_prefix, src_case_suffix, cases):
    g.g += f"{enum_type_name} {fn_name}({src_type_name} tag) {{\n"
    g.g += f"\tswitch (tag) {{ \n"
    assert type(cases) is list
    for node in cases:
        name = node["name"]
        g.g += f"\t\tcase {src_case_prefix}{name}{src_case_suffix}: \n"
        clazz = node.get("class")
        match(clazz):
            case None:
                g.g += "\t\t\treturn 0;\n"
                pass
            case _ if type(clazz) is str:
                cap = capitalize(clazz)
                g.g += f"\t\t\treturn {enum_case_prefix}{cap};\n"
                pass
            case _ if type(clazz) is list:
                g.g += "\t\t\treturn "
                for (j, clazzj) in zip(range(0, len(clazz)), clazz):
                    if j > 0:
                        g.g += " | "
                    cap = capitalize(clazzj)
                    g.g += f"{enum_case_prefix}{cap}"
                g.g += ";\n"
    g.g += f"\t\tdefault: assert(false);\n"
    g.g += f"\t}}\n"
    g.g += f"\tSHADY_UNREACHABLE;\n"
    g.g += f"}}\n"